### PR TITLE
Add `raven check` for full workspace diagnostics in CI; unify package-library init

### PR DESCRIPTION
**Pull request by jbearak**
## What

Adds a new `raven check` CLI subcommand and refactors the package-library initialization path to remove editor/CI drift.

### `raven check` — full diagnostics for CI
- New subcommand that indexes the workspace and reports the **full** diagnostic set (syntax, semantic, style lints, cross-file, package, and undefined-variable) — the same diagnostics the editor publishes, run headlessly. This is the CI counterpart to `raven lint`, which only runs style rules in isolation.
- Auto-detects R (honoring `packages.enabled` / `packages.rPath` / `packages.additionalLibraryPaths`) to resolve installed-package exports and base symbols; degrades gracefully with a one-line stderr note when R is absent.
- Exit codes match `raven lint` (`0` ok, `1` threshold breach / usage error, `2` operator error).
- Docs: new `docs/cli.md#raven-check` section, README/configuration updates, GitHub Actions + SARIF example.

### Shared CLI code (`cli/shared.rs`)
- Extracts `OutputFormat`, `SeverityLevel`, the `text`/`json`/`sarif` renderers, severity gating, file-type predicates, and the recursive R-file walker out of `cli/lint.rs` so `check`, `lint`, and `analysis-stats` share one implementation instead of each other's internals.

### Shared `build_package_library` helper (kills editor/CI drift)
- The "build a `PackageLibrary` from config" sequence was hand-reproduced in **four** places that had already drifted (readiness computed before vs after `add_library_paths`, inconsistent `spawn_blocking` use). All four now route through one `build_package_library` helper in `package_library.rs`.
- Readiness + degradation precedence live in a single pure `classify` function with a table-tested truth table, making the illegal "ready, but degraded" state unrepresentable via a `{ library, status }` outcome + `PackageLibraryStatus` enum.
- Adjacent `packages_enabled` gates cleaned up: the redundant settings-reload gate is collapsed (the helper self-gates); the load-bearing `refreshPackages` gate is kept with a comment + regression test (it must not clobber an existing library when packages are disabled).
- `find_package_directory` now requires NAMESPACE/DESCRIPTION to be a readable **file**, not just an existing path — fixes acceptance of a directory named `NAMESPACE`.

## Why

`raven lint` parses files in isolation, so CI couldn't gate on cross-file, undefined-variable, or package diagnostics. `raven check` closes that gap. The shared helper exists because CI diagnostics must match the editor's, and the four parallel init sites were the structural cause of drift (one already shipped without the `packages.enabled` gate).

## Reviewer notes

- The design rationale (four sites, the readiness predicate, the dead-but-kept `InitFailed` variant, why `get_fallback_lib_paths` makes some branches unreachable end-to-end) is documented in `docs/superpowers/specs/2026-05-29-package-library-helper-design.md`.
- `cli/check.rs` re-sequences shared building blocks (`recompute_parsed_configs` → `scan_workspace` → `apply_workspace_index` → `initialize_package_inputs_from_state` → `build_package_library`) rather than calling a single LSP orchestrator, because the LSP startup is bound to locks/client/background tasks and isn't extractable. The drift-prone part (package-library construction) is the part that *was* centralized.
- The interesting `classify` branches are unreachable on supported platforms (the cfg fallback is always non-empty), so they're pinned by the pure classifier's truth table rather than end-to-end tests.

## Tests
- `cargo test -p raven --lib package_library::` — 79 passed.
- New tests: `classify` truth table, `build_package_library` disabled/additional-paths, the two adjacent-gate regressions (settings-reload-disabled, refresh-while-disabled), `raven check` parse/collect/exit-code coverage, symlink-skipping walker, and the unusable-package-dir skip.

## Summary by CodeRabbit

* **New Features**
  * Added a headless "raven check" subcommand for full workspace diagnostics; shares text/json/sarif renderers with lint and honors severity thresholds.

* **Bug Fixes / Reliability**
  * More robust package-awareness and initialization (preserves existing library when packages disabled), safer workspace scanning, and consistent handling of unreadable or non-UTF encodings.

* **Documentation**
  * README and CLI docs clarified differences between `raven check` (full diagnostics) and `raven lint` (style-only) and documented usage/exit codes.

* **Tests**
  * Added/expanded tests covering CLI args, package gates, file decoding, and config discovery.

[Review Change Stack](https://app.coderabbit.ai/change-stack/jbearak/raven/pull/342?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: crates/raven/src/backend.rs [hunk 1]
```diff
@@ -2474,6 +2453,13 @@ impl LanguageServer for Backend {
...
                 let packages_enabled =
                     { self.state.read().await.cross_file_config.packages_enabled };
+                // Load-bearing gate (do not collapse into `rebuild_package_library`
+                // the way the settings-reload site does): here the swap into
+                // `state` happens *inside* this `if`. Routing through the helper
+                // unconditionally would swap in `new_empty()` when packages are
+                // disabled, clobbering the library the user built before
+                // disabling — a `refreshPackages` issued while disabled must
+                // leave the existing library untouched.
                 if packages_enabled {
```
[the remark this answers is not in the document]
Comment (jbearak) [reply crates/raven/src/backend.rs]: Good catch — fixed in c02bc70d. `raven.refreshPackages` now early-returns when package awareness is disabled, so the cache clear, help-cache flush, prefetch, libpath-watcher restart, and republish are all skipped (not just the swap). Strengthened `refresh_packages_disabled_keeps_existing_library` to seed a cache entry and assert it survives, which the old code would have wiped.

File: docs/superpowers/specs/2026-05-29-package-library-helper-design.md [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+- **`backend.rs:5719` removal:** add a regression test that a settings reload with `packages_enabled == false` yields `package_library` empty and `package_library_ready == false` (locking in that the now-unconditional `rebuild_package_library` call preserves the disabled outcome). Confirm the existing settings-reload tests stay green.
+- **`backend.rs:2476` (kept):** verify (assert if not already covered) that `refreshPackages` while disabled does **not** clear an already-populated library — the behavior the load-bearing gate protects.
+
+### Verification commands (run from the worktree, NOT the main checkout)
+```
+cargo test -p raven --lib package_library::   # helper + classify tests
+cargo test -p raven --lib cli::               # CLI module
+cargo test -p raven --lib maybe_init_r        # gate/additional-paths tests
+cargo test -p raven --lib rebuild_package     # backend builder tests
+```
```
[the remark this answers is not in the document]
Comment (jbearak) [reply docs/superpowers/specs/2026-05-29-package-library-helper-design.md]: Done in 040e25ae — tagged the fence as `bash`.

File: crates/raven/src/state.rs [hunk 3]
```diff
@@ -1184,17 +1215,97 @@ fn collect_file_paths_inner(dir: &Path, out: &mut Vec<PathBuf>, visited: &mut Ha
...
+        let pair = [c[0], c[1]];
+        if little_endian {
+            u16::from_le_bytes(pair)
+        } else {
+            u16::from_be_bytes(pair)
+        }
+    });
+    char::decode_utf16(units)
+        .collect::<Result<String, _>>()
+        .map_err(|_| SourceReadError::InvalidEncoding { offset: 0, byte: 0 })
```
[the remark this answers is not in the document]
Comment (jbearak) [reply crates/raven/src/state.rs:1326]: Fixed in 3f4906cc — `decode_utf16` now returns `InvalidEncoding` for odd-length input instead of letting `chunks_exact` drop the trailing byte and accept truncated input. Added regression test `decode_source_rejects_truncated_utf16`.

File: crates/raven/src/cli/lint.rs [hunk 4]
```diff
@@ -134,80 +92,101 @@ Exit codes:
...
+                // every per-file `[[linting.overrides]]` patch (same failure
+                // mode as commit 81978f0 fixed for the non-explicit root).
+                let parent = explicit.parent().unwrap_or(cwd).to_path_buf();
                 let root = if parent.is_absolute() {
                     parent
                 } else {
                     cwd.join(&parent)
                 };
-                (root, Some(l.settings), false)
+                Ok((root, Some(l.settings), false))
```
[the remark this answers is not in the document]
Comment (jbearak) [reply crates/raven/src/cli/lint.rs:152]: Fixed in 3f4906cc — `root` is now lexically normalized via `normalize_path_public` (resolves `.`/`..` without touching the filesystem) so `strip_prefix(root)` keeps matching files under a `--config ../pkg/raven.toml` root. Added regression test `resolve_lint_config_normalizes_explicit_config_root`. (Note: a fully robust fix for *relative* file args carrying `..` would also need the file side normalized at the comparison point, which lives in shared LSP code; out of scope for this minimal change.)

File: crates/raven/src/cli/check.rs [hunk 5]
```diff
@@ -0,0 +1,911 @@
...
+/// decoded into lines; the byte offset in the message is the actionable
+/// pointer. `byte == 0` marks the rare malformed-UTF-16 case, where there's no
+/// single offending byte to name.
+fn encoding_diagnostic(offset: usize, byte: u8) -> Diagnostic {
+    use tower_lsp::lsp_types::{DiagnosticSeverity, Position, Range};
+    let detail = if byte == 0 {
+        // No single offending byte to name: a truncated multibyte sequence at
+        // EOF, or malformed/odd-length UTF-16 (which carried a BOM). Keep this
+        // encoding-agnostic so it doesn't misattribute either case.
+        "could not be decoded as UTF-8 or UTF-16".to_string()
```
[the remark this answers is not in the document]
Comment (jbearak) [reply crates/raven/src/cli/check.rs]: Fixed in a603844f — `encoding_diagnostic` now composes the full message per branch instead of prefixing a fixed "File is not valid UTF-8:". The `byte == 0` case (truncated UTF-8 at EOF, or malformed UTF-16 that carried a BOM) now reads "File could not be decoded as UTF-8 or UTF-16. …"; the Latin-1 case keeps the precise byte/offset wording. Added unit test `encoding_diagnostic_message_matches_the_failure_kind`.
